File: 2025-season6/9주차/최영권/MergeKSortedLists.java
package LinkedList;

import LinkedList.Node.ListNode;

import java.util.PriorityQueue;

/**
 * 시간복잡도 : O(Nlogk) , N은 전체 노드의 총 개수 , k: 리스트의 개수
 */
public class MergeKSortedLists {
    public static void main(String[] args) {
        ListNode head1 = new ListNode(1, new ListNode(4, new ListNode(5)));
        ListNode head2 = new ListNode(1, new ListNode(3, new ListNode(4)));
        ListNode head3 = new ListNode(2, new ListNode(6));
        ListNode[] lists = new ListNode[]{head1, head2, head3};
        MergeKSortedLists mergeKSortedLists = new MergeKSortedLists();
        ListNode listNode = mergeKSortedLists.mergeKLists(lists);
        System.out.println(listNode);
    }

    public ListNode mergeKLists(ListNode[] lists) {
        int k = lists.length;
        if (k == 0) {
            return null;
        }

        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a, b) -> a.val - b.val);

        // 최소힙에 가장 작은수(맨앞) 들을 넣음
        for (ListNode list : lists) {
            if (list != null) {
                minHeap.add(list);
            }
        }

        ListNode result = new ListNode();
        ListNode first = result;

        // 힙이 비어있지 않으면, 반복하면서 가장 작은수를 빼서 결과 리스트노드에 넣고, 다음 노드를 다시 최소힙에 넣는 형태
        while (!minHeap.isEmpty()) {
            ListNode node = minHeap.poll();
            result.next = node;
            result = result.next;
            if (node.next != null) {
                minHeap.add(node.next);
            }
        }

        return first.next;
    }
}

File: 2025-season6/9주차/merge-k-sorted-lists.kt
class Solution {
    // 전체 시간복잡도: O(N log k)
    // 전체 공간복잡도: O(k)
    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        // 우선순위 큐 초기화 - O(k)
        val pq = PriorityQueue<ListNode>(compareBy { it.`val` })
        for (node in lists) {
            if (node != null) pq.add(node) // O(log k)
        }

        // dummy 노드로 결과 리스트 구성 - O(1)
        val dummy = ListNode(0)
        var tail = dummy

        // 큐가 빌 때까지 최소 노드를 꺼내며 병합 - O(N log k)
        while (pq.isNotEmpty()) {
            val minNode = pq.poll() // O(log k)
            tail.next = minNode
            tail = tail.next!!
            if (minNode.next != null) pq.add(minNode.next) // O(log k)
        }

        // 결과 반환 - O(1)
        return dummy.next
    }
}
